src/index.ts

```diff
@@ -153,7 +156,29 @@ function* lexer(str: string): Generator<LexToken, LexToken> {
       throw new TypeError(`Missing parameter name at ${i}: ${DEBUG_URL}`);
     }
 
-    return value;
+    if (chars[i] === "(" && options?.pattern) {
```

Comment:
Take a look at the previous code here: https://github.com/pillarjs/path-to-regexp/commit/60f2121e9b66b7b622cc01080df0aabda9eedee6#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80L155

I'd use that as a starting point, e.g. move the `(` into the `lexer` itself (not part of `name`) and yield a `PATTERN` directly. Handle the fact that `PATTERN` is expected in a certain position in the `if (param)` below by doing `tryConsume("PATTERN")`.